rs/types/base_types/src/canister_id.rs

```diff
@@ -90,6 +93,43 @@ impl CanisterId {
 
         Self(PrincipalId::new_opaque_from_array(data, blob_length))
     }
+
+    /// Converts from PrincipalId.
+    ///
+    /// The problem with CanisterId::try_from(principal_id) is that it lies.
+    //
+    // Keep this consistent with from_u64.
+    pub fn try_from_principal_id(principal_id: PrincipalId) -> Result<Self, CanisterIdError> {
+        // Must be opaque.
+        if principal_id.class() != Ok(PrincipalIdClass::Opaque) {
+            return Err(CanisterIdError::InvalidPrincipalId(format!(
+                "Principal ID {} is of class {:?} (not Opaque).",
+                principal_id,
+                principal_id.class(),
+            )));
+        }
+
+        // Must be of length 10.
+        let raw = principal_id.as_slice();
+        if raw.len() != 10 {
+            return Err(CanisterIdError::InvalidPrincipalId(format!(
+                "Principal ID {} consists of {} bytes (not 10).",
+                principal_id,
+                raw.len(),
+            )));
+        }
+
+        // Byte 8 (penultimate) must be 0x01.
+        if raw[8] != 0x01 {
```

Comment:
I'm not sure this is a check that can be considered conclusive. If it's opaque and the penultimate byte is 0x01 then you know it's a principal that was created by the IC and it could be a canister id, but I think it could also be a subnet id or node id. Or at the very least you have no guarantee that it's a canister id only.

I don't think it changes much for what you're trying to do here though -- we can still proceed with this as it does improve current state in the sense that if you want to handle errors or expect errors when trying to convert from a principal to a potential canister id, then this is now an actual option.

Reply:
Correct, my comment was purely informational. I agree that if we want to fix the issue that `from_u64` can return potentially non-canister IDs, it should be addressed separately.